### PR TITLE
Feature Flags Basic Implementation

### PR DESCRIPTION
**Pull request by mlaws21**
Basic implementation of feature flags and UI to use them, I implemented them for the about, wiki, and FAQ pages, there is also infrastructure to implement the feature flags for nav elements too. Big picture you add flags to components/FeatureFlagsConfig and then can edit them in the admin dashboard.

Sorry for the big PR, I got a bit carried away

### CODE REVIEW

File: src/components/FeatureFlags.tsx [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import React from "react";
+import { features } from "./FeatureFlagsConfig";
+
+// TODO: Resets to default on fail?
+
+export const fetchFeatures = async () => {
+  return Promise.resolve({ features });
+};
```
Comment (yechs) [inline src/components/FeatureFlags.tsx:8]: Do we need this to be async?

File: src/index.js [hunk 2]
```diff
@@ -22,14 +22,19 @@ import history from "./lib/history";
...
 
 ReactDOM.render(
   <React.StrictMode>
     <Provider store={store}>
       <PersistGate loading={null} persistor={persistor}>
         <HistoryRouter history={history}>
-          <App />
+          <FeatureFlagsProvider>
+            <App />
+          </FeatureFlagsProvider>
```
Comment (yechs) [inline src/index.js:37]: I don't think writing your own context is a good idea. By default, React context is not persistent through page refreshes.

However, we do use Redux, and I have Redux state persistence infrastructure. A better idea would be storing/accessing the feature flags in Redux

File: src/components/FeatureFlags.tsx
Comment (yechs) [inline src/components/FeatureFlags.tsx:1]: See my comment on using custom React Context vs Redux

**Review by yechs — CHANGES_REQUESTED** [review]
TL;DR: Use Redux instead of React Context

Redux is complicated and there are many ways of writing them. We should talk tomorrow or some other time to introduce you to them

**Comment by yechs** [comment]
However, if you are curious about how Redux current works in our implementation. You might want to look at these files. At this point, I might as well write a short doc on it.

## HowTo: work with Modern Redux states in WSO-React

### Key Concepts

The official Redux tutorial is also helpful. Read the following article to make sure you understand these terms: reducer, action, dispatch, selectors.
https://redux.js.org/tutorials/essentials/part-1-overview-concepts

### Make sure to read modern Redux (using Redux Toolkit, or RTK)

If you are searching for external documentation/tutorial. Make sure you are reading the modern Redux codes (with Redux Toolkit) instead of the old Redux (using React-Redux, which requires lots of boilerplates). Specifically, we are adopting the Slices style. Read more:
https://redux.js.org/tutorials/fundamentals/part-7-standard-patterns
https://redux.js.org/tutorials/fundamentals/part-8-modern-redux

### Reducer and Action definition

* `src/lib/authSlice.ts` defines a slice of reducers and actions:
https://github.com/WilliamsStudentsOnline/wso-react/blob/447ca378b1ac73b8e16838f355d564acc389e0f9/src/lib/authSlice.ts
* The corresponding `authReducer` is then re-exported in `src/reducers/index.ts`
https://github.com/WilliamsStudentsOnline/wso-react/blob/447ca378b1ac73b8e16838f355d564acc389e0f9/src/reducers/index.ts

### Redux store configuration & Persistence

The Redux store is configured in `src/lib/store.ts`. 
https://github.com/WilliamsStudentsOnline/wso-react/blob/447ca378b1ac73b8e16838f355d564acc389e0f9/src/lib/store.ts#L27-L32
In the file we also persist the reducer so its states are persisted through page refreshes.
https://github.com/WilliamsStudentsOnline/wso-react/blob/447ca378b1ac73b8e16838f355d564acc389e0f9/src/lib/store.ts#L17-L20
The file creates two React Hooks (`useAppDispatch` and `useAppSelector`) to be used in actual pages
https://github.com/WilliamsStudentsOnline/wso-react/blob/447ca378b1ac73b8e16838f355d564acc389e0f9/src/lib/store.ts#L60-L65

### Usage Example

An example usage is in `App.tsx`
Read (with selector):
https://github.com/WilliamsStudentsOnline/wso-react/blob/447ca378b1ac73b8e16838f355d564acc389e0f9/src/components/App.tsx#L52-L56
Write (by dispatching actions):
https://github.com/WilliamsStudentsOnline/wso-react/blob/447ca378b1ac73b8e16838f355d564acc389e0f9/src/components/App.tsx#L75-L82

File: src/index.js (hunk 2 above)
Comment (yechs) [reply src/index.js:37]: Also, BTW, Redux (and the redux state persistor we use) is built upon React Contexts. So essentially you are just switching to a different Context and getting many perks for free.

https://github.com/WilliamsStudentsOnline/wso-react/blob/4635ca3c39e90d2136f4a87ff132adee6b810c60/src/index.js#L13-L15
https://github.com/WilliamsStudentsOnline/wso-react/blob/4635ca3c39e90d2136f4a87ff132adee6b810c60/src/index.js#L32-L33

File: src/lib/featureFlagSlice.ts [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  reducers: {
+    toggleFeatureFlag: (
+      state,
+      action: PayloadAction<keyof FeatureFlagsState>
+    ) => {
+      const flagKey = action.payload;
+      if (flagKey in state) {
+        state[flagKey] = !state[flagKey];
+      }
+    },
```
Comment (yechs) [inline src/lib/featureFlagSlice.ts:33]: I'm a little worried about toggling, because it is not idempotent (and also because redux is async). 
Maybe the action should take in both `flagKey` and `newState`?

File: src/lib/featureFlagSlice.ts [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+  // Add more feature flags as needed
+}
+
+export const initialState: FeatureFlagsState = {
+  enableAbout: false,
+  enableWiki: false,
+  enableFAQ: false,
+
+  // Initialize other feature flags
+};
```
Comment (yechs) [inline src/lib/featureFlagSlice.ts:19]: A design choice to be had: should these be enabled by default? And then only disabled if the API says so...

File: src/reducers/index.ts [hunk 5]
```diff
@@ -1,14 +1,31 @@
...
 };
 
+export interface RootState {
+  courseState: ReturnType<typeof courseReducer>;
+  schedulerUtilState: ReturnType<typeof schedulerUtilReducer>;
+  authState: ReturnType<typeof authReducer>;
+  goodrichState: ReturnType<typeof goodrichReducer>;
+  featureFlagState: ReturnType<typeof featureFlagReducer>;
+}
+
```
Comment (yechs) [inline src/reducers/index.ts:23]: You should not write this manually. They are auto generated here

https://github.com/WilliamsStudentsOnline/wso-react/blob/d3e763402a11a6e41a81b3737741c3df6077eb7f/src/lib/store.ts#L71

File: src/components/Nav.tsx [hunk 6]
```diff
@@ -11,6 +13,30 @@ import { Link } from "react-router-dom";
...
+  flag: keyof RootState["featureFlagState"];
+}
+
+const FeatureFlagElement: React.FC<FeatureFlagElementProps> = ({
+  element,
+  flag,
+}) => {
+  const enabled = useSelector(
+    (state: RootState) => state.featureFlagState[flag]
+  );
```
Comment (yechs) [inline src/components/Nav.tsx:31]: You should use these hooks and they are automatically typed

https://github.com/WilliamsStudentsOnline/wso-react/blob/d3e763402a11a6e41a81b3737741c3df6077eb7f/src/lib/store.ts#L74-L76

File: src/components/App.tsx [hunk 7]
```diff
@@ -194,15 +196,20 @@ const App = () => {
...
             }
           />
+
           <Route path="schedulecourses" element={<Scheduler />} />
           {/* Static Content Pages */}
+
           <Route path="about" element={<About />} />
           <Route path="faq" element={<FAQ />} />
           <Route path="team" element={<Team />} />
+          <Route path="admin" element={<Dashboard />} />
```
Comment (yechs) [inline src/components/App.tsx:206]: After our discussion today, I think this dashboard should only appear to users with `ScopeAdminAll`.

Use the `<RequireScope>` component to help you. and configure the scope requirements here:

https://github.com/WilliamsStudentsOnline/wso-react/blob/56390d43b2cd9bb6caec69d1e612f9f94f577815/src/router-permissions.js#L13-L38

**Review by yechs — CHANGES_REQUESTED** [review]
TL;DR: 
- [x] Redux hooks and typing
- [ ] scope protect the admin page

File: src/lib/featureFlagSlice.ts (hunk 3 above)
Comment (mlaws21) [reply src/lib/featureFlagSlice.ts:33]: I don't understand the idea here, the values can only be "on" or "off", why does toggling not make sense?

File: src/lib/featureFlagSlice.ts (hunk 4 above)
Comment (mlaws21) [reply src/lib/featureFlagSlice.ts:19]: Sure, that makes sense

**Comment by mlaws21** [comment]
all changes requested changes are implemented besides changing the toggle, I don't understand exactly why that change would be necessary. A given feature should be on or off, and there is nothing to change to besides the other?

**Comment by yechs** [comment]
> A given feature should be on or off, and there is nothing to change to besides the other?

True, but the toggle action is not idempotent. 

I did some more search and it looks like redux actions are not strictly required to be idempotent. 
However, writing it in an idempotent way makes your code more robust, especially in asynchronous contexts.
For example, you may dispatch two toggle actions, or two actions that set the same value to true.
The result would be different.

Also imagine the case where the data is corrupted (you never know what the browser/user would do to your stored data...).
Having it set to true/false instead of negating it would be more robust against unforeseen conditions.

**Comment by AndrewMuh** [comment]
I would also add that boolean states are generally bad practice. It's best to store Enum states for clarity and in case we would like to support other feature states. For example we might want to support a senior-only state or a student-only state.

File: src/components/App.tsx [hunk 8]
```diff
@@ -194,15 +196,27 @@ const App = () => {
...
           <Route path="schedulecourses" element={<Scheduler />} />
           {/* Static Content Pages */}
+
           <Route path="about" element={<About />} />
           <Route path="faq" element={<FAQ />} />
           <Route path="team" element={<Team />} />
+          <Route
+            path="admin"
+            element={
+              <RequireScope token={apiToken} name="admin.dashboard">
```
Comment (yechs) [inline src/components/App.tsx:209]: This route name should be just `admin` (since the path name is also just `/admin`)

File: src/router-permissions.js [hunk 9]
```diff
@@ -35,6 +35,10 @@ const routePermissions = {
     tokenLevel: 3,
     scopes: [scopes.ScopeGoodrich, scopes.ScopeGoodrichManager],
   },
+  "admin.dashboard": {
+    tokenLevel: 3,
+    scopes: [scopes.ScopeAdminAll],
+  },
```
Comment (yechs) [inline src/router-permissions.js:41]: Same as above, change route name to `admin`

File: src/components/views/Misc/About.jsx [hunk 10]
```diff
@@ -118,4 +119,19 @@ const About = () => {
...
 };
 
+const About = () => {
+  const enableAbout = useAppSelector(
+    (state) => state.featureFlagState["enableAbout"]
+  );
+
+  return (
+    <div>
+      {enableAbout === "Enabled" ? (
```
Comment (yechs) [inline src/components/views/Misc/About.jsx:129]: Don't use magic keywords `"Enabled"` here. Use `FFState.Enabled` instead. Same in `FAQ.jsx` and in `Nav.tsx` (`FeatureFlagElement`)

File: src/lib/featureFlagSlice.ts [hunk 11]
```diff
@@ -0,0 +1,45 @@
+import { createSlice, PayloadAction } from "@reduxjs/toolkit";
+
+// const API_ADDRESS = "http://localhost:8080";
+
+export enum FFState {
+  Enabled = "Enabled",
+  Disabled = "Disabled",
+  Pending = "Pending",
+}
+
```
Comment (yechs) [inline src/lib/featureFlagSlice.ts:10]: Regarding typing here, I think we need some more discussions.

@AndrewMuh, do you think it's a better idea if we give each flag its own type for granular state mangament?
If so, should we implement it in union types? (given that enum cannot be extended) Or totally separate types?

Also, do you think the rendering logic like `senior-only` should be handled at backend or front-end?

This question is important, because it will affect how `FeatureFlagElement` should be implemented (in Nav.tsx)

**Review by yechs — CHANGES_REQUESTED** [review]
To Change:
- [ ] Route name
- [ ] avoid magic keywords

Need Discussion:
- [ ] typing of states
- [ ] frontend/backend logic divide

File: src/lib/featureFlagSlice.ts (hunk 11 above)
Comment (AndrewMuh) [reply src/lib/featureFlagSlice.ts:10]: Honestly thinking about it now, I think all flags should be either Enabled or Disabled (I still think enums, but only these two). Things like senior only or student only should be done through a "featureIsSeniorOnly" feature flag. This simplifies implementation in the frontend/db: just check ephmatchEnabled and ephmatchIsSeniorOnly for example.

File: src/components/Nav.tsx [hunk 12]
```diff
@@ -11,6 +13,33 @@ import { Link } from "react-router-dom";
...
+  element,
+  flag,
+}) => {
+  const enabled = useAppSelector(
+    (state: RootState) => state.featureFlagState[flag]
+  );
+
+  return (
+    <div className="nav_feature_div">
+      {console.log(enabled)}
```
Comment (AndrewMuh) [inline src/components/Nav.tsx:35]: delete

File: src/components/Nav.tsx [hunk 13]
```diff
@@ -111,18 +140,42 @@ const Nav = () => {
...
+            <FeatureFlagElement
+              element={<a href="/wiki/">Wiki</a>}
+              flag="enableWiki"
+            />
+            <FeatureFlagElement
+              element={<Link to="about">About</Link>}
+              flag="enableAbout"
+            />
+
+            {/* {featureFlagElement(
```
Comment (AndrewMuh) [inline src/components/Nav.tsx:156]: delete commented code

File: src/lib/featureFlagSlice.ts [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+  enableAbout: FFState.Enabled,
+  enableWiki: FFState.Enabled,
+  enableFAQ: FFState.Enabled,
+};
+
+const featureFlagSlice = createSlice({
+  name: "featureFlags",
+  initialState,
+  reducers: {
+    toggleFeatureFlag: (
```
Comment (AndrewMuh) [inline src/lib/featureFlagSlice.ts:28]: Rename to setFeatureFlag?

File: src/lib/featureFlagSlice.ts [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+      if (flag in state) {
+        state[flag] = value;
+      }
+    },
+  },
+});
+
+export const { toggleFeatureFlag } = featureFlagSlice.actions;
+
+// export const getFeatureFlag = (state: FeatureFlagsState, flag: keyof FeatureFlagsState) => featureFlagsState[flag];
```
Comment (AndrewMuh) [inline src/lib/featureFlagSlice.ts:42]: remove commented code

**Review by AndrewMuh — CHANGES_REQUESTED** [review]
fix small comments

File: src/lib/featureFlagSlice.ts (hunk 15 above)
Comment (yechs) [reply src/lib/featureFlagSlice.ts:42]: This hasn't been removed yet

File: src/lib/featureFlagSlice.ts [hunk 16]
```diff
@@ -0,0 +1,44 @@
+import { createSlice, PayloadAction } from "@reduxjs/toolkit";
+
+// const API_ADDRESS = "http://localhost:8080";
```
Comment (yechs) [inline src/lib/featureFlagSlice.ts:3]: Also here

**Review by yechs — CHANGES_REQUESTED** [review]

**Comment by mlaws21** [comment]
I added some styling to dashboard.tsx -- I probably should have made a new pr but I didn't want to branch off of another pr

**Review by yechs — COMMENTED** [review]
So far so good. 

Some more code will need to be added on backend integration
- [ ] Admin Dashboard: when flags are changed, send request to update them in backend API server
- [ ] All pages (perhaps in App.tsx, similar to `updateAPI()`): on page load, send request to backend API for current feature flags

Related backend PR: https://github.com/WilliamsStudentsOnline/wso-go/pull/201
